docs/DesignDocs/RemoteAttestationCollaterals.md

```diff
@@ -0,0 +1,292 @@
+    0,
+    NULL, // opt_params
+    0,
+    &report,
+    report_size);
+...
+```
+
+##### Challenger verifies the evidence with custom collaterals (in the untrusted host)
+```C
+...
+//
+// Challenger gets collaterals not using OE SDK
+//
+
+//
+// Challenger builds **collaterals** structure
+//
+
+// Verify report with collaterals
+result = oe_verify_remote_report_with_collaterals(
+            report,
+            report_size,
+            collaterals,
+            collaterals_size,
+            NULL, // input_validation_time
+            NULL, // parsed_report);
+...
+```
+
+Specification
+-------------
+
+### Collateral structure
+Generic serializable structure that stores the collaterals in raw binary format.
+
+`report.h`
+```C
+/*! \struct oe_collaterals_t
+ *
+ * \brief OE collaterals
+ *
+ * Raw generic serializable structure that contains the collaterals.
+ *
+ */
+typedef struct _oe_collaterals_t
+{
+    uint32_t version;
+    uint8_t enclave_type;         ///< The type of enclave
+    uint8_t reserved;
+    uint16_t reserved2;
+    uint32_t num_elements;
+    size_t buffer_size;           ///< Size of the buffer
+
+    ///< Data buffer is made of an offset array of type uint32_t, followed by
+    ///< the actual data.
+    ///< This array has the size of **num_elements** and stores the offset
+    ///< into the data section.
+    ///<  _________________________
+    ///<  |  version              |
+    ///<  |-----------------------|
+    ///<  |  reserved    | type   |
+    ///<  |-----------------------|
+    ///<  |  num_elements         |
+    ///<  |-----------------------|
+    ///<  |  buffer_size          |
+    ///<  |-----------------------|
+    ///<  |                       |
+    ///<  |  array(uint32_t)      |
+    ///<  |  size of num_elements |
+    ///<  |-----------------------|
+    ///<  |  Data                 |
+    ///<  |_______________________|
+    ///<
+    uint8_t buffer[];              ///< Buffer of offsets + data
+
+} oe_collaterals_t;
+```
+
+### New Collateral functions
+
+`collaterals.c`
+```C
+/*! \struct sgx_collaterals
+ *
+ * \brief SGX collateral structure
+ *
+ * The generic oe_collaterals_t structure is parsed and converted into this
+ * internal structure.  The order of the generic data elements should
+ * coincide with the order of the fields in this structure.
+ */
+typedef struct _sgx_collaterals_t
+{
+    uint8_t* tcb_info;                      ///< TCB info
+    size_t tcb_info_size;                   ///< TCB Info size
+    uint8_t* tcb_issuer_chain;              ///< PEM format
+    size_t tcb_issuer_chain_size;           ///< Size of the tcb_issuer_chain
+    uint8_t* crl[3];                        ///< CRLs
+    size_t crl_size[3];                     ///< CRLs sizes
+    uint8_t* crl_issuer_chain[3];           ///< PEM format
+    size_t crl_issuer_chain_size[3];        ///< Size of each crl_issuer_chain
+
+    uint8_t* qe_id_info;                    ///< QE Identity info
```

Comment:
Is the QE identity info specific to the machine or machine-independent?

I'd like to make sure that the same "collaterals" struct doesn't try to combine machine-specific and machine-independent info into the same struct, especially if this struct is sent across a network.  That's so (for example) machine specific information can be sent by an Attester's machine, whereas machine-independent info could be obtained separately (or cached across many machines) by a Verifier.   The "num_elements" design above should allow multiple independent structs to be used together in the same message just fine, if I understand correctly, so there's no need to use a combined struct.

Reply:
Yes it should allow multiple independent structs.  With the plug-in model, you will be able to define your own set of attestation functions and define your endorsement data format.

Reply:
"QE identity info" in the context of endorsement here is not machine dependent